android-components/components/support/ktx/src/main/java/mozilla/components/support/ktx/android/view/Activity.kt

```diff
@@ -19,20 +18,24 @@ import mozilla.components.support.base.log.logger.Logger
 /**
  * Attempts to enter immersive mode - fullscreen with the status bar and navigation buttons hidden,
  * expending itself into the notch area for devices running API 28+.
- * This will automatically register and use an
- * - an inset listener: [View.OnApplyWindowInsetsListener] on API 30+ or
- * - a system visibility listener: [View.OnSystemUiVisibilityChangeListener] for below APIs.
  *
+ * This will automatically register and use an inset listener: [View.OnApplyWindowInsetsListener]
  * to restore immersive mode if interactions with various other widgets like the keyboard or dialogs
  * got the activity out of immersive mode without [exitImmersiveMode] being called.
  */
-fun Activity.enterToImmersiveMode() {
-    setAsImmersive()
-    enableImmersiveModeRestore()
-}
+fun Activity.enterImmersiveMode(
+    insetsController: WindowInsetsControllerCompat = window.createWindowInsetsController(),
+) {
+    hideInsets(insetsController)
+
+    ViewCompat.setOnApplyWindowInsetsListener(window.decorView) { view, insetsCompat ->
+        if (insetsCompat.isVisible(WindowInsetsCompat.Type.statusBars())) {
+            hideInsets(insetsController)
+        }
+        // Allow the decor view to have a chance to process the incoming WindowInsets.
+        onApplyWindowInsets(view, insetsCompat)
+    }
```

Comment:
nit: Was this move needed? Personally preferred this being in the separate fun, it made it self evident why the block was required

Reply:
I have a bit of bias against putting everything in a separate small function that is used just in one place :) it's a "shallow vs deep functions" discourse, and it's a pretty obvious listener. Also, `exitImmersiveMode` does have any inner functions. I wonder to put all of the logic inside the enter/exit functions, except for `hideInsets` that's used by the listener as well.